### PR TITLE
[B-03724] Ghost Websocket

### PR DESCRIPTION
**Pull request by freesig**
:ghost: :ghost: :ghost: 
Ready to be reviewed.
Extras:
- [x] Update conductor api to breaking change [here](https://github.com/holochain/holochain-conductor-api/pull/48)
- [x] Write bench on new ghost actor 
- [x] Write bench on websocket

### CODE REVIEW

**Comment by freesig** [comment]
Overly simplified benchmark of various parts of this crate
image

File: crates/holochain_websocket/examples/nodejs_echo_client/nodejs_echo_client.js [hunk 1]
```diff
@@ -29,9 +29,11 @@ class HolochainWebsocket {
   }
```
Comment (neonphog) [inline crates/holochain_websocket/examples/nodejs_echo_client/nodejs_echo_client.js:29]: This file actually was supposed to be executable, note the shebang at the top : )

Comment (freesig) [reply crates/holochain_websocket/examples/nodejs_echo_client/nodejs_echo_client.js:29]: Not exactly sure what you mean. This was just copied from the old verison

Comment (neonphog) [reply crates/holochain_websocket/examples/nodejs_echo_client/nodejs_echo_client.js:29]: oh, git shows the executable bit changing: `100755 → 100644`

Comment (neonphog) [reply crates/holochain_websocket/examples/nodejs_echo_client/nodejs_echo_client.js:29]: zippy has a setup that occasionally *adds* the executable bit to random files like `lib.rs`... there is another file you moved that removed one of these, so I thought you were doing it on purpose : )

File: crates/holochain_websocket/src/websocket.rs [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+        // - Shutdown from_socket task because to_socket task has shutdown.
+        let (shutdown_from_socket, from_socket) = Valved::new(from_socket);
+        // - Shutdown from_socket task because both channel ends have dropped.
+        // PairShutdown will also send a close message to to_socket.
+        let from_socket = pair_shutdown.wrap(from_socket);
+        // - Shutdown to_socket task because from_socket task has shutdown.
+        // This valve will not close is to_socket can successfully send a close message to from_socket.
+        let (shutdown_to_socket, to_websocket) = Valved::new(to_websocket);
+
+        // Spawn the to external task.
```
Comment (neonphog) [inline crates/holochain_websocket/src/websocket.rs:236]:
```suggestion
        // Spawn the "to" external task.
```

File: crates/holochain_websocket/src/websocket.rs [hunk 3]
```diff
@@ -0,0 +1,763 @@
...
+        tokio::task::spawn(
+            self.clone()
+                .run_to_socket(to_socket, to_websocket, shutdown_from_socket)
+                .in_current_span(),
+        );
+        // Spawn the from external task.
```
Comment (neonphog) [inline crates/holochain_websocket/src/websocket.rs:242]:
```suggestion
        // Spawn the "from" external task.
```

File: crates/holochain_websocket/src/websocket.rs [hunk 4]
```diff
@@ -0,0 +1,763 @@
...
+/// Actor that tracks responses.
+pub struct Websocket(GhostActor<WebsocketInner>);
+
+#[derive(Debug)]
+struct ResponseTracker {
+    /// List of responses where Some is an registered response
+    /// while None is a free slot.
+    responses: Vec<Option<RegisterResponse>>,
+    /// Queue of free slots to avoid ever growing memory.
+    free_indices: VecDeque<usize>,
```
Comment (neonphog) [inline crates/holochain_websocket/src/websocket.rs:46]: Consider a slow remote. We timeout a response / re-use an id. Now that previous response comes in, and is matched with the wrong request.

HashMaps are pretty efficient these days, I'd recommend using an always incrementing `u64` for message id, and storing the responses in a HashMap.

File: crates/holochain_websocket/src/websocket_sender.rs [hunk 5]
```diff
@@ -1,150 +1,187 @@
...
-
-            recv.await.map_err(|e| Error::new(ErrorKind::Other, e))?;
-
-            Ok(())
+    #[tracing::instrument(skip(self))]
+    /// Make a request to for the other side to respond to.
+    ///
+    /// Note:
+    /// There is no timeouts in this code. You either need to wrap
+    /// this future in a timeout or use [`WebsocketSender::request_timeout`].
```
Comment (neonphog) [inline crates/holochain_websocket/src/websocket_sender.rs:109]:
```suggestion
```

This note was meant for the fn below.

File: crates/holochain/src/conductor/interface/websocket.rs [hunk 6]
```diff
@@ -120,106 +98,71 @@ pub async fn spawn_app_interface_task<A: InterfaceApi>(
...
             Err(e) => {
                 error!(error = &e as &dyn std::error::Error)
             }
             Ok(()) => {}
         }
     }
+    // Do an atomic checked sub.
+    // This can still fail to decrement but won't overflow.
+    // This is ok because we really only need a rough idea if of the number of connections
+    // and failing to decrement should be rare.
```
Comment (neonphog) [inline crates/holochain/src/conductor/interface/websocket.rs:157]: could also use an atomicisize and let it go negative : )

File: crates/holochain_websocket/src/websocket.rs (hunk 4 above)
Comment (neonphog) [reply crates/holochain_websocket/src/websocket.rs:46]: Actually - I don't see any code to prune stale responses. We'll probably need to add that too.

**Review by neonphog — CHANGES_REQUESTED** [review]
This is amazing, looks awesome to me. Just need the couple mentioned fixes to the response handling.

File: crates/holochain_websocket/examples/nodejs_echo_client/nodejs_echo_client.js (hunk 1 above)
Comment (freesig) [reply crates/holochain_websocket/examples/nodejs_echo_client/nodejs_echo_client.js:29]: Oh I see what you mean now

**Review by neonphog — APPROVED** [review]

**Pull request merged**